kernel/posix/pthread_rwlock.c

```diff
@@ -0,0 +1,304 @@
+	rwlock->blocked_reader_count = 0;
+	rwlock->owner = NULL;
+	rwlock->status = INITIALIZED;
+	return 0;
+}
+
+/**
+ * @brief Destroy read-write lock object.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_rwlock_destroy(pthread_rwlock_t *rwlock)
+{
+	if (rwlock->status == NOT_INITIALIZED) {
+		return EINVAL;
+	}
+
+	if (rwlock->owner != NULL) {
+		return EBUSY;
+	}
+
+	if (rwlock->status == INITIALIZED) {
+		rwlock->status = NOT_INITIALIZED;
+		return 0;
+	}
+
+	return EINVAL;
+}
+
+/**
+ * @brief Lock a read-write lock object for reading.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_rwlock_rdlock(pthread_rwlock_t *rwlock)
+{
+	if (rwlock->status == NOT_INITIALIZED) {
+		return EINVAL;
+	}
+
+	return read_lock_acquire(rwlock, K_FOREVER);
+}
+
+/**
+ * @brief Lock a read-write lock object for reading within specific time.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_rwlock_timedrdlock(pthread_rwlock_t *rwlock,
+			       const struct timespec *abstime)
+{
+	u32_t  timeout;
+
+	if (rwlock->status == NOT_INITIALIZED || abstime->tv_nsec < 0 ||
+	    abstime->tv_nsec > NSEC_PER_SEC) {
+		return EINVAL;
+	}
+
+	timeout = calculate_timeout(abstime);
+	return read_lock_acquire(rwlock, timeout);
+
+}
+
+/**
+ * @brief Lock a read-write lock object for reading immedately.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_rwlock_tryrdlock(pthread_rwlock_t *rwlock)
+{
+	if (rwlock->status == NOT_INITIALIZED) {
+		return EINVAL;
+	}
+
+	return read_lock_acquire(rwlock, K_NO_WAIT);
+}
+
+/**
+ * @brief Lock a read-write lock object for writing.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_rwlock_wrlock(pthread_rwlock_t *rwlock)
+{
+	if (rwlock->status == NOT_INITIALIZED) {
+		return EINVAL;
+	}
+
+	k_sem_take(&rwlock->sem, K_FOREVER);
+	rwlock->owner = k_current_get();
+
+	return 0;
+}
+
+/**
+ * @brief Lock a read-write lock object for writing within specific time.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_rwlock_timedwrlock(pthread_rwlock_t *rwlock,
+			       const struct timespec *abstime)
+{
+	u32_t timeout, ret = 0;
+
+	if (rwlock->status == NOT_INITIALIZED || abstime->tv_nsec < 0 ||
+	    abstime->tv_nsec > NSEC_PER_SEC) {
+		return EINVAL;
+	}
+
+	timeout = calculate_timeout(abstime);
+
+	if (k_sem_take(&rwlock->sem, timeout) == 0) {
+		rwlock->owner = k_current_get();
+	} else {
+		ret = ETIMEDOUT;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Lock a read-write lock object for writing immedately.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_rwlock_trywrlock(pthread_rwlock_t *rwlock)
+{
+	if (rwlock->status == NOT_INITIALIZED) {
+		return EINVAL;
+	}
+
+	if (k_sem_take(&rwlock->sem, K_NO_WAIT) == 0) {
+		rwlock->owner = k_current_get();
+		return 0;
+	}
+
+	return EBUSY;
+}
+
+/**
+ *
+ * @brief Unlock a read-write lock object.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_rwlock_unlock(pthread_rwlock_t *rwlock)
+{
+	s32_t count;
+
+	if (rwlock->status == NOT_INITIALIZED) {
+		return EINVAL;
+	}
+
+	k_sem_take(&rwlock->count_lock, K_FOREVER);
+
+	if (rwlock->current_reader_count != 0) {
```

Comment:
This might become moot if you end up changing rwlock to be implemented with two semaphores, but if  current_reader_count drops to a negative number, weird things will happen.

Reply:
Yes if we change implement to separate  semaphores for RD and WR, this becomes a topic of debate,  But in current implementation it is needed. 
current_reader_count will not become less than 1 as at a time only one thread can write into it as it is protected through I am protecting update of this through "count_lock" semaphore.